### ISSUE TITLE
Walls behind slopes gets drawn on top of them

### ISSUE THREAD

**Issue by Broxzier** [OpenRCT2/OpenRCT2#2239]
The same bug occurs in the original game.

![Screenshot](http://i.imgur.com/wlcG1t1.png)

**Comment by Christipher** [comment]
Can affirm that this bug occurs in the original game as well... and very easy to reproduce as well...

**Comment by Paulicus25** [comment]
Downloaded new nightly version 22nd March 2025. Bug still present. 100% reproducible every time. If NOT on stable sort, can be worked around by placing a scenery item on top of sloped land in front of wall, like a small bush...if so the wall behind no draws properly behind the wall. However if stable sort is turned on to improve other issues elsewhere in the park, this workaround of placing a bush in front to make it draw correctly no longer functions.

In short, turning stable sort in makes this problem impossible to work around. If stable sort is planned to be an eventual replacement to the current default, then solving this issue increases in priority. Especially with multiple other drawing issues involving walls and layering currently being added to the code, yet this thoroughly irritating issue many years old remaining.

**Comment by mixiate** [comment]
Yeah, if there is going to be an alternative sort, it really needs to fix this and other fundamental issues with the sort algorithm. It doesn't just affect walls but other things too. This can't be fixed by default because it will break many existing parks.

I'm not sure yet what is going to be the best way to fix all these issues. I posted this in the discord a while back, but just for the sake of it here's my explanation of why this happens:

Sprites get put in to 16 pixel high rows (or quadrants as they are called in the code), based on their bounding boxes x and y values summed. If x + y >= 32 then a sprite gets put in to the next quadrant.

https://github.com/OpenRCT2/OpenRCT2/blob/65b5d86e6ccc5ad371a2fb826b277faf10b1303d/src/openrct2/paint/Paint.cpp#L65-L104

Front wall bounding boxes have x and y values of 4 and 28, causing them to go to the next quadrant compared to the majority of other sprites on a tile. That's why making them 2 and 28 for example prevents this from happening.

When sprites are painted they are added to the front of the linked list for that quadrant, so the last sprite to be painted ends up first. If you draw the unsorted list it will be mostly backwards.

The sort checks the bounding boxes and if it is in front of another, it swaps them.

When a wall is painted in the background tile, and gets put in the quadrant with the foreground tile, it will be the first or one of the first things painted in that quadrant, so before the sort it will be at the end of the list.

Because its bounding box is not in front of the land bounding box, it doesn't get swapped with it. So it stays after the land tile, and gets drawn over it.

In this picture these sprites are all in the single quadrant in red, even though they are placed on different tiles.

![Image](https://github.com/user-attachments/assets/10221072-f53b-4ed5-b152-6e6b518e21ae)

**Comment by mixiate** [comment]
Yeah it's possible to fix. Putting other things on the tile causes the bounding boxes to sort in a different order. It's not really a fix but a quirk of the sorting algorithm, which as I said is flawed in the first place to cause this.

You can even fix just the walls themselves by making sure their bounding box x + y isn't 32 or more, something that is done for many other things.

But the reason why you can't just fix it is because there is 20+ years of parks that rely on this glitch. It's actually the reason why you can build with walls in zero clearances and it mostly works how you would expect. It would need to be part of an alternative sorting algorithm, which needs many other considerations.
